wrong site URL detected

While behind the proxy, Metabase is picking up internal server IP as site URL while it should do that for the public IP

```
2017-03-26T03:23:22.351849507Z 03-26 03:23:22 INFO metabase.core :: Metabase Initialization COMPLETE
2017-03-26T03:23:22.777886392Z 03-26 03:23:22 INFO metabase.middleware :: Setting Metabase site URL to 10.128.0.6
```

Infrastructure: GCE, Kubernetes 1.6

Not sure if this is happening in other docker based environments.

I am experiencing a similar Issue after upgrading from 0.32.5 to 0.32.8.
The log is populated with `Invalid site URL` messages upon each https request but everything seems to work fine.

Metabase is running on a webserver behind a reverse-proxy (I changed it's original FQDN to `reverseproxy.internal.domain`) that can be accessed via https only. No http.
In the logs you can see that metabase tries to set the **correct** URL in the first place but fails to validate it, because that URL **is not reachable from the webserver's perspective**.

Setting the URL `https://reverseproxy.internal.domain/metabase` manually inside the admin panel also fails with message `Assert failed: Invalid site URL: https://reverseproxy.internal.domain/metabase (u/url? s)`

```
05-16 11:08:25 ERROR metabase.public-settings :: site-url is invalid; returning nil for now. Will be reset on next request.
java.lang.AssertionError: Assert failed: Invalid site URL: https://reverseproxy.internal.domain/metabase
(u/url? s)
        at metabase.public_settings$normalize_site_url.invokeStatic(public_settings.clj:55)
        at metabase.public_settings$normalize_site_url.invoke(public_settings.clj:47)
        at metabase.public_settings$fn__18233$fn__18237.invoke(public_settings.clj:65)
        at metabase.models.setting$get.invokeStatic(setting.clj:211)
        at metabase.models.setting$get.invoke(setting.clj:205)
        at metabase.models.setting$setting_fn$fn__17623.invoke(setting.clj:434)
        at metabase.middleware.misc$maybe_set_site_url_STAR_.invokeStatic(misc.clj:45)
        at metabase.middleware.misc$maybe_set_site_url_STAR_.invoke(misc.clj:43)
        at metabase.middleware.misc$maybe_set_site_url$fn__61523.invoke(misc.clj:57)
        at metabase.middleware.misc$bind_user_locale$fn__61526.invoke(misc.clj:74)
        at ring.middleware.cookies$wrap_cookies$fn__61663.invoke(cookies.clj:177)
        at metabase.middleware.misc$add_content_type$fn__61511.invoke(misc.clj:30)
        at metabase.middleware.misc$disable_streaming_buffering$fn__61548.invoke(misc.clj:119)
        at metabase.middleware.misc$wrap_gzip$fn__61541.invoke(misc.clj:101)
        at ring.adapter.jetty$async_proxy_handler$fn__61259.invoke(jetty.clj:35)
        at ring.adapter.jetty.proxy$org.eclipse.jetty.server.handler.AbstractHandler$ff19274a.handle(Unknown Source)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
        at org.eclipse.jetty.server.Server.handle(Server.java:502)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:370)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:267)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
        at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
        at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:765)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:683)
        at java.lang.Thread.run(Thread.java:748)
05-16 11:08:25 INFO middleware.misc :: Setze Metabase-URL auf reverseproxy.internal.domain
05-16 11:08:25 WARN middleware.misc :: Failed to set site-url
java.lang.AssertionError: Assert failed: Invalid site URL: http://reverseproxy.internal.domain
(u/url? s)
        at metabase.public_settings$normalize_site_url.invokeStatic(public_settings.clj:55)
        at metabase.public_settings$normalize_site_url.invoke(public_settings.clj:47)
        at metabase.public_settings$fn__18233$fn__18234.invoke(public_settings.clj:69)
        at metabase.models.setting$set_BANG_.invokeStatic(setting.clj:368)
        at metabase.models.setting$set_BANG_.invoke(setting.clj:356)
        at metabase.models.setting$setting_fn$fn__17623.invoke(setting.clj:440)
        at metabase.middleware.misc$maybe_set_site_url_STAR_.invokeStatic(misc.clj:49)
        at metabase.middleware.misc$maybe_set_site_url_STAR_.invoke(misc.clj:43)
        at metabase.middleware.misc$maybe_set_site_url$fn__61523.invoke(misc.clj:57)
        at metabase.middleware.misc$bind_user_locale$fn__61526.invoke(misc.clj:74)
        at ring.middleware.cookies$wrap_cookies$fn__61663.invoke(cookies.clj:177)
        at metabase.middleware.misc$add_content_type$fn__61511.invoke(misc.clj:30)
        at metabase.middleware.misc$disable_streaming_buffering$fn__61548.invoke(misc.clj:119)
        at metabase.middleware.misc$wrap_gzip$fn__61541.invoke(misc.clj:101)
        at ring.adapter.jetty$async_proxy_handler$fn__61259.invoke(jetty.clj:35)
        at ring.adapter.jetty.proxy$org.eclipse.jetty.server.handler.AbstractHandler$ff19274a.handle(Unknown Source)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
        at org.eclipse.jetty.server.Server.handle(Server.java:502)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:370)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:267)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
        at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
        at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:765)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:683)
        at java.lang.Thread.run(Thread.java:748)
```

--- a/src/metabase/util.clj
+++ b/src/metabase/util.clj
@@ -25,8 +25,7 @@
            java.util.concurrent.TimeoutException
            java.util.Locale
            javax.xml.bind.DatatypeConverter
-           org.apache.commons.validator.routines.UrlValidator
-           org.apache.commons.validator.routines.RegexValidator))
+           [org.apache.commons.validator.routines RegexValidator UrlValidator]))
 
 ;; This is the very first log message that will get printed.
 ;;

--- a/src/metabase/util.clj
+++ b/src/metabase/util.clj
@@ -25,7 +25,8 @@
            java.util.concurrent.TimeoutException
            java.util.Locale
            javax.xml.bind.DatatypeConverter
-           org.apache.commons.validator.routines.UrlValidator))
+           org.apache.commons.validator.routines.UrlValidator
+           org.apache.commons.validator.routines.RegexValidator))
 
 ;; This is the very first log message that will get printed.
 ;;

--- a/src/metabase/util.clj
+++ b/src/metabase/util.clj
@@ -100,7 +100,9 @@
 (defn url?
   "Is `s` a valid HTTP/HTTPS URL string?"
   ^Boolean [s]
-  (let [validator (UrlValidator. (varargs String ["http" "https"]) UrlValidator/ALLOW_LOCAL_URLS)]
+  (let [validator (UrlValidator. (varargs String ["http" "https"])
+                                 (RegexValidator. "^\\p{Alnum}+([\\.|\\-]\\p{Alnum}+)*(:\\d*)?")
+                                 UrlValidator/ALLOW_LOCAL_URLS)]
     (.isValid validator (str s))))
 
 (defn maybe?

--- a/test/metabase/public_settings_test.clj
+++ b/test/metabase/public_settings_test.clj
@@ -1,7 +1,6 @@
 (ns metabase.public-settings-test
   (:require [clojure.test :refer :all]
             [environ.core :as env]
-            [expectations :refer [expect]]
             [metabase.models.setting :as setting]
             [metabase.public-settings :as public-settings]
             [metabase.test
@@ -13,95 +12,88 @@
 
 (use-fixtures :once (fixtures/initialize :db))
 
- ;; double-check that setting the `site-url` setting will automatically strip off trailing slashes
-(expect
-  "http://localhost:3000"
-  (tu/with-temporary-setting-values [site-url nil]
-    (public-settings/site-url "http://localhost:3000/")
-    (public-settings/site-url)))
+;; double-check that setting the `site-url` setting will automatically strip off trailing slashes
+(deftest site-url-settings
+  (is (= "http://localhost:3000"
+         (tu/with-temporary-setting-values [site-url nil]
+           (public-settings/site-url "http://localhost:3000/")
+           (public-settings/site-url)))))
 
- ;; double-check that setting the `site-url` setting will prepend `http://` if no protocol was specified
-(expect
-  "http://localhost:3000"
-  (tu/with-temporary-setting-values [site-url nil]
-    (public-settings/site-url "localhost:3000")
-    (public-settings/site-url)))
+;; double-check that setting the `site-url` setting will prepend `http://` if no protocol was specified
+(deftest site-url-settings-prepend-http
+  (is (= "http://localhost:3000"
+         (tu/with-temporary-setting-values [site-url nil]
+           (public-settings/site-url "localhost:3000")
+           (public-settings/site-url)))))
 
-(expect
-  "http://localhost:3000"
-  (tu/with-temporary-setting-values [site-url nil]
-    (public-settings/site-url "localhost:3000")
-    (public-settings/site-url)))
-
-(expect
-  "http://localhost:3000"
-  (tu/with-temporary-setting-values [site-url nil]
-    (public-settings/site-url "http://localhost:3000")
-    (public-settings/site-url)))
+(deftest site-url-settings-with-no-trailing-slash
+  (is (= "http://localhost:3000"
+         (tu/with-temporary-setting-values [site-url nil]
+           (public-settings/site-url "http://localhost:3000")
+           (public-settings/site-url)))))
 
 ;; if https:// was specified it should keep it
-(expect
-  "https://localhost:3000"
-  (tu/with-temporary-setting-values [site-url nil]
-    (public-settings/site-url "https://localhost:3000")
-    (public-settings/site-url)))
+(deftest site-url-settings-https
+  (is (= "https://localhost:3000"
+         (tu/with-temporary-setting-values [site-url nil]
+           (public-settings/site-url "https://localhost:3000")
+           (public-settings/site-url)))))
 
 ;; we should not be allowed to set an invalid `site-url` (#9850)
-(expect
-  AssertionError
-  (tu/with-temporary-setting-values [site-url nil]
-    (public-settings/site-url "http://https://www.camsaul.com")))
+(deftest site-url-settings-validate-site-url
+  (is (thrown? AssertionError
+               (tu/with-temporary-setting-values [site-url nil]
+                 (public-settings/site-url "http://https://www.camsaul.com")))))
 
-(expect
-  AssertionError
-  (tu/with-temporary-setting-values [site-url nil]
-    (public-settings/site-url "https://www.camsaul.x")))
+(deftest site-url-settings-set-valid-domain-name
+  (is (tu/with-temporary-setting-values [site-url nil]
+        (public-settings/site-url "https://www.camsaul.x"))))
 
 ;; if `site-url` in the database is invalid, the getter for `site-url` should return `nil` (#9849)
-(expect
-  {:get-string "https://www.camsaul.x", :site-url nil}
-  (tu.log/suppress-output
-    (tu/with-temporary-setting-values [site-url "https://metabase.com"]
-      (setting/set-string! :site-url "https://www.camsaul.x")
-      {:get-string (setting/get-string :site-url)
-       :site-url   (public-settings/site-url)})))
+(deftest site-url-settings-nil-getter-when-invalid
+  (is (= {:get-string "https://&", :site-url nil}
+         (tu.log/suppress-output
+          (tu/with-temporary-setting-values [site-url "https://metabase.com"]
+            (setting/set-string! :site-url "https://&")
+            {:get-string (setting/get-string :site-url)
+             :site-url   (public-settings/site-url)})))))
 
 ;; We should normalize `site-url` when set via env var we should still normalize it (#9764)
-(expect
-  {:get-string "localhost:3000/", :site-url "http://localhost:3000"}
-  (with-redefs [env/env (assoc env/env :mb-site-url "localhost:3000/")]
-    (tu/with-temporary-setting-values [site-url nil]
-      {:get-string (setting/get-string :site-url)
-       :site-url   (public-settings/site-url)})))
+(deftest site-url-settings-normalize
+  (is (= {:get-string "localhost:3000/", :site-url "http://localhost:3000"}
+         (with-redefs [env/env (assoc env/env :mb-site-url "localhost:3000/")]
+           (tu/with-temporary-setting-values [site-url nil]
+             {:get-string (setting/get-string :site-url)
+              :site-url   (public-settings/site-url)})))))
 
 (deftest invalid-site-url-env-var-test
   {:get-string "asd_12w31%$;", :site-url nil}
   (testing (str "If `site-url` is set via an env var, and it's invalid, we should return `nil` rather than having the"
                 " whole instance break")
     (tu.log/suppress-output
-      (with-redefs [env/env (assoc env/env :mb-site-url "asd_12w31%$;")]
-        (tu/with-temporary-setting-values [site-url nil]
-          (is (= "asd_12w31%$;"
-                 (setting/get-string :site-url)))
-          (is (= nil
-                 (public-settings/site-url))))))))
+     (with-redefs [env/env (assoc env/env :mb-site-url "asd_12w31%$;")]
+       (tu/with-temporary-setting-values [site-url nil]
+         (is (= "asd_12w31%$;"
+                (setting/get-string :site-url)))
+         (is (= nil
+                (public-settings/site-url))))))))
 
-(expect
-  "HOST"
-  (let [zz (i18n/string-as-locale "zz")]
-    (i18n/with-user-locale zz
-      (str (:display-name (first (get-in (public-settings/public-settings) [:engines :postgres :details-fields])))))))
+(deftest translate-public-setting
+  (is (= "HOST"
+         (let [zz (i18n/string-as-locale "zz")]
+           (i18n/with-user-locale zz
+             (str (:display-name (first (get-in (public-settings/public-settings) [:engines :postgres :details-fields])))))))))
 
-(expect
-  [true "HOST"]
-  (let [zz (i18n/string-as-locale "zz")]
-    (i18n/with-user-locale zz
-      [(= zz (i18n/user-locale))
-       (tru "Host")])))
+(deftest tru-translates
+  (is (= [true "HOST"]
+         (let [zz (i18n/string-as-locale "zz")]
+           (i18n/with-user-locale zz
+             [(= zz (i18n/user-locale))
+              (tru "Host")])))))
 
 ;; Make sure Max Cache Entry Size can be set via with a string value, which is what comes back from the API (#9143)
-(expect
-  "1000"
-  ;; use with temp value macro so original value gets reset after test run
-  (tu/with-temporary-setting-values [query-caching-max-kb nil]
-    (public-settings/query-caching-max-kb "1000")))
+(deftest max-cache-entry
+  (is (= "1000"
+         ;; use with temp value macro so original value gets reset after test run
+         (tu/with-temporary-setting-values [query-caching-max-kb nil]
+           (public-settings/query-caching-max-kb "1000")))))

--- a/test/metabase/util_test.clj
+++ b/test/metabase/util_test.clj
@@ -32,6 +32,7 @@
 (expect true  (u/url? "http://www.cool.com:3000"))
 (expect true  (u/url? "http://localhost:3000/auth/reset_password/144_f98987de-53ca-4335-81da-31bb0de8ea2b#new"))
 (expect true  (u/url? "http://192.168.1.10/"))
+(expect true  (u/url? "http://metabase.intranet/"))
 (expect false (u/url? "google.com"))                      ; missing protocol
 (expect false (u/url? "ftp://metabase.com"))              ; protocol isn't HTTP/HTTPS
 (expect false (u/url? "http://.com"))                     ; no domain

--- a/test/metabase/util_test.clj
+++ b/test/metabase/util_test.clj
@@ -37,6 +37,7 @@
 (expect false (u/url? "ftp://metabase.com"))              ; protocol isn't HTTP/HTTPS
 (expect false (u/url? "http://.com"))                     ; no domain
 (expect false (u/url? "http://google."))                  ; no TLD
+(expect false (u/url? "http://"))                         ; no domain or tld
 (expect false (u/url? "http:/"))                          ; nil .getAuthority needs to be handled or NullPointerException
 
 

